sudo_plugin/src/output/conv_facility.rs

```diff
@@ -0,0 +1,165 @@
+//! A module that implements the conversations API for a sudo plugin
+
+use crate::sys;
+
+use std::sync::{Arc, Mutex};
+use sudo_plugin_sys::{sudo_conv_t, sudo_conv_message, sudo_conv_reply};
+use std::io;
+use std::ffi::CString;
+use std::mem;
+use std::ptr;
+use std::slice;
+
+/// `ConvMsgType` is the type of conversation prompt as specified by 
+/// the sudo plugin
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum ConvMsgType {
+    /// Do not echo user input
+    PromptEchoOff  = sys::SUDO_CONV_PROMPT_ECHO_OFF,
+    /// Echo user input
+    PromptEchoOn   = sys::SUDO_CONV_PROMPT_ECHO_ON,
+    /// The prompt is an Error Message
+    ErrorMsg       = sys::SUDO_CONV_ERROR_MSG,
+    /// The prompt is an informational message
+    InfoMsg        = sys::SUDO_CONV_INFO_MSG,
+    /// Mask user input
+    PrompMask      = sys::SUDO_CONV_PROMPT_MASK,
+    /// Allows for echo if no TTY
+    PromptEchoOk   = sys::SUDO_CONV_PROMPT_ECHO_OK,
+    // This is only available on plugin version 1.14 which isn't supported yet
+    //ConvPreferTTY      = sys::SUDO_CONV_PREFER_TTY, /* flag: use tty if possible */
+}
+
+/// `ConversationPrompt` is the struct that holds the actual prompt displayed
+/// to the user. 
+#[derive(Clone, Debug)]
+pub struct ConversationPrompt {
+    /// The type of prompt
+    pub msg_type: ConvMsgType,
+    /// The timeout for the prompt. 0 is no timeout
+    pub timeout: i32,
+    /// The message to be displayed
+    pub msg: String
+}
+
+impl ConversationPrompt {
+    /// This is an internal method for converting a `ConversationPrompt` to the
+    /// `sudo_conv_message` type for FFI
+    fn convert_to_conv_message(&self) -> io::Result<sudo_conv_message> {
+        // TODO: can I get rid of this clone?
+        let message = CString::new(self.msg.clone()).map_err(|err|
+            io::Error::new(io::ErrorKind::Other, err)
+        )?;
+
+        Ok(sudo_conv_message {
+            msg_type: self.msg_type as i32,
+            timeout: self.timeout,
+            // msg: message.as_ptr()
+            msg: message.into_raw()
+        })
+    } 
+}
+
+/// `ConversationReply` is the reply (if any) from the user to our promt
+#[derive(Clone, Debug)]
+pub struct ConversationReply {
+    /// The reply by the user
+    pub reply: String
+}
+
+impl ConversationReply {
+    /// Internal method for converting `sudo_conv_reply` to `Option<ConversationReply>` to expose only safe APIs
+    fn from_conv_reply(scr: sudo_conv_reply) -> Option<ConversationReply> {
+        if scr.reply.is_null() {
+            return None;
+        }
+        unsafe { 
+            Some( ConversationReply {
+                reply: CString::from_raw(scr.reply).into_string().expect("error converting reply to String")
```

Comment:
[CString::from_raw](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_raw) is only intended to take ownership of a value created with `CString::to_raw`. I expect this could be a problem if a custom allocator was in use.

Reply:
Do you have a different recommendation for de-referencing a `*c_char` to a `CString` over FFI?

Reply:
looking at the src to `from_raw` it looks like this wouldn't be an issue so long as the string being passed in is a null terminated string, which it would be coming from sudo.

https://doc.rust-lang.org/src/std/ffi/c_str.rs.html#475-486

Reply:
> Do you have a different recommendation for de-referencing a *c_char to a CString over FFI?

I think you should copy the data to a newly allocated string via `CStr`, then deallocate the provided reply using the system allocator explicitly.

> looking at the src to from_raw it looks like this wouldn't be an issue so long as the string being passed in is a null terminated string, which it would be coming from sudo.

It's not just about the memory layout - memory given to us from sudo is allocated using the system allocator, but CString expects its memory to have been allocated using the global rust allocator, which aren't necessarily the same.

Reply:
Agreed with @steven-joruk on this. `CStr` is meant to be a wrapper around non-owned NULL-terminated C-strings, like those you might receive from a library over FFI.

Reply:
Also I've obviously completely dropped the ball on merging this. I'll make sure that happens this week.

Reply:
Sorry I lost track of this, here's an example of copying the data and using the system deallocator, in case jemalloc or some other custom global allocator is in use by Rust:

```rust
let reply = unsafe { CStr::from_ptr(scr.reply) }.to_str().unwrap().to_owned();
unsafe { libc::free(scr.reply as *mut c_void) };
Some(ConversationReply { reply })
```

Unwrapping isn't ideal, of course.

I appreciate this has been a long time LivingInSyn, if you're busy would it be okay for me to update this with the code review requests, some tests and a change along these lines? I'd like to see this PR get merged in the next couple of months, it'll be very useful.

Reply:
I'm more than happy to accept a PR for this, but since the underlying API changed underneath this, a lot more work needs to be done to merge this PR into main